[POS][Custom payment UI] – Switch to Payment Controller in POS Totals | Base PR

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->
#### Switch to Payment Controller in WooPosTotalsViewModel for payment collection

💡 This is a base PR that includes the following partial PRs:
1. https://github.com/woocommerce/woocommerce-android/pull/13037
2. https://github.com/woocommerce/woocommerce-android/pull/13038
3. https://github.com/woocommerce/woocommerce-android/pull/13053
4. https://github.com/woocommerce/woocommerce-android/pull/13057
5. https://github.com/woocommerce/woocommerce-android/pull/13090
6. https://github.com/woocommerce/woocommerce-android/pull/13100

⚠️ Don't merge until all the partial PRs above are marked merged.

Closes: #12822
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR is first step toward integrating `CardReaderPaymentController` into `WooPosTotals` screen.

In addition to using `CardReaderPaymentController` into `WooPosTotalsViewModel`, an old code specific to old IPP flow reuse in POS was removed.

In the next PRs:
1. Handling cancelations (e.g. on back press)
8. Removing "Collect Payment" button
9. Showing "Connect to a reader" error in Totals screen
10. Introducing new Totals screen UI states with animations for payment states (loading, collecting, capturing, processing)
11. Showing payment errors

💡 The PR targets a feature branch `feature/custom-payment-ui`. "Do not merge" label added to prevent merging at the moment so that we are able to release the full feature.

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
Test that it's possible to collect payment (happy path) and that "Payment success" screen is shown after payment is completed.

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
1. Reader connect/disconnect using floating menu
2. (POS) payment collection happy path

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Remove custom `if (isPos)` from controller

This allows observing all the states in POS, without the dialog-based UI. Previously custom `if` statement was used to avoid showing (duplicate) success dialog.

samiuelson committed:
Display `paymentState`s in Totals pane in POS

samiuelson committed:
Remove POS payment failure handling

POS-specific payment failure handling is not applicable, because we're now observing payments states using `CardReaderPaymentController.paymentState`. In the dialog-based Payment flow, the result payment result is passed here — to the SelectPaymentMethod screen (start of the payment flow nav graph).

samiuelson committed:
Return result to POS before opening Payment dialog

samiuelson committed:
Pass reader connection result (success/failure) to POS

The new result is designed both for reader "connection" and "prepare for payment" actions.

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>TTPPaymentProgressDelegate</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

samiuelson committed:
Throw `IllegalStateException` if POS uses old IPP flow

samiuelson committed:
Update tests in `WooPosTotalsViewModelTest`

samiuelson committed:
Satisfy detekt's complaints

samiuelson committed:
Start POS reader connection process in `ReaderStatusChecker`

* There is no need to open `SelectPaymentMethod` we can start directly in `CardReaderStatusChecker` both for reader connection and preparation for payment.

samiuelson committed:
Remove POS-specific code in IPP that's unused now

samiuelson committed:
Remove POS-specific unavailable code

samiuelson committed:
Rename reader status to `WooPosCardReaderConnectionStatus`

samiuelson committed:
Get rid of `WooPosCardReaderFacade.collectPayment()` flow

The only requirement for reader is to be connected. There's no need to pass any other intent to the IPP flow. Once reader connects we can engage Payment Controller to collect payment.

samiuelson committed:
Get rid of `WooPosCardReaderFacade.collectPayment()` flow

The only requirement for reader is to be connected. There's no need to pass any other intent to the IPP flow. Once reader connects we can engage Payment Controller to collect payment.

samiuelson committed:
Remove redundant code

samiuelson committed:
Remove redundant code

samiuelson committed:
Simplify POS event names and models in IPP

samiuelson committed:
Remove redundant code

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/payment/controller/CardReaderPaymentController.kt:
```diff
@@ -459,20 +455,10 @@ class CardReaderPaymentController(
     ) {
         paymentReceiptHelper.storeReceiptUrl(orderId, paymentStatus.receiptUrl)
         appPrefs.setCardReaderSuccessfulPaymentTime()
-        if (paymentOrRefund.isPOS) {
```
💡 Previously this `if` was preventing from showing "payment successful" dialog in order to show success screen in POS only. This is no longer needed, because we are observing paymentState directly in `WooPosTotalsViewModel`.

samiuelson committed:
Remove unused import

samiuelson committed:
Simplify code

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/methodselection/SelectPaymentMethodEvent.kt:
```diff
@@ -23,11 +23,6 @@ data class NavigateToCardReaderPaymentFlow(
     val cardReaderType: CardReaderType
 ) : MultiLiveEvent.Event()
 
-data class SkipScreenInPosAndNavigateToCardReaderPaymentFlow(
```
💡 `SelectPaymentMethod` screen is not open anymore. When connecting to a reader we open `CardReaderStatusChecker` directly.

samiuelson committed:
Remove no more accessed code

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/cardreader/WooPosCardReaderActivity.kt:
```diff
@@ -37,26 +35,6 @@ class WooPosCardReaderActivity : AppCompatActivity(R.layout.activity_woo_pos_car
         ) { requestKey, bundle ->
             when (requestKey) {
                 WOO_POS_CARD_PAYMENT_REQUEST_KEY -> {
-                    val result = bundle.parcelable<WooPosCardReaderPaymentStatus>(
```
💡 We don't need to read the payment result now, because we're only calling IPP flow in order to connect to a reader. Once we receive the result from IPP flow we need to finish `WooPosCardReaderActivity`.

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/cardreader/WooPosCardReaderFacade.kt:
```diff
@@ -1,54 +1,29 @@
-        WooPosCardReaderPaymentStatus.Unknown
-    )
-    val paymentStatus: Flow<WooPosCardReaderPaymentStatus> = _paymentStatus
+    val readerStatus: StateFlow<CardReaderStatus> = cardReaderManager.readerStatus
```
💡 We don't need any more data than `CardReaderStatus` now.

kidinov replied:
Didn't get this one. Why it can not keep being flow?

samiuelson replied:
When using payment controller now in Totals, we need to read `readerStatus.value` to decide whether to go to "payment collect" state or show "Reader not connected" state based on reader connection status. That's why we need a hot and stateful flow, instead of cold and stateless. The underlying `CardReaderManager.readerStatus` is of the `StateFlow` type as well.

kidinov replied:
Oh I see you need this for `cardReaderFacade.readerStatus.value is Connected`

kidinov replied:
However, I am not sure if we'll need to use it like that. We may want to use `collect` and depending on that show "connect to your reader" state or start the payment right away

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsScreen.kt:
```diff
@@ -138,6 +138,11 @@ private fun TotalsLoaded(
 
             Spacer(modifier = Modifier.weight(1f))
+
+            Text(
```
💡 Temporary debug Text. Will be replaced in next PRs.

samiuelson committed:
Remove unused code

samiuelson committed:
Merge branch `custom-payment-ui-5` into `custom-payment-ui-6-use-payment-controller-in-pos`

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/connect/CardReaderConnectViewModel.kt:
```diff
@@ -527,25 +524,23 @@ class CardReaderConnectViewModel @Inject constructor(
+        val param = arguments.cardReaderFlowParam
+        if (param is CardReaderFlowParam.WooPosConnection) {
+            returnToWooPos()
+        } else if (!connected) {
```
Previously, `returnToWooPos` was called only when not connected. Now, it is also invoked when connected. Is this expected?

kidinov replied:
If is then we may code the same slightly more nicely

```
private fun exitFlow(connected: Boolean) {
    val param = arguments.cardReaderFlowParam
    when {
        param is CardReaderFlowParam.WooPosConnection -> returnToWooPos()
        !connected -> triggerEvent(ExitWithResult(false))
        else -> triggerEvent(
            ShowCardReaderTutorial(
                arguments.cardReaderFlowParam,
                arguments.cardReaderType
            )
        )
    }
}

```

samiuelson replied:
> Previously, returnToWooPos was called only when not connected. Now, it is also invoked when connected. Is this expected?

It was a bug. We should return result to POS always not only in a happy path.

> we may code the same slightly more nicely

Done 👍  – 41e778bd209ab4d64e59c26fbdbbdc3321c00511

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/statuschecker/CardReaderStatusCheckerViewModel.kt:
```diff
@@ -127,5 +127,7 @@ class CardReaderStatusCheckerViewModel
             val cardReaderType: CardReaderType,
         ) : MultiLiveEvent.Event()
+
+        data object ReturnToWooPos : MultiLiveEvent.Event()
```
That's never triggered. Expected?

samiuelson replied:
Good catch. This was added here by mistake. Reverted – e67149f6a072ed12a9c1e0cc0ab595f8cf02ceab

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/methodselection/SelectPaymentMethodViewModel.kt:
```diff
@@ -120,7 +120,7 @@ class SelectPaymentMethodViewModel @Inject constructor(
                             when (param.paymentType) {
                                 SIMPLE, ORDER, ORDER_CREATION, TRY_TAP_TO_PAY -> showPaymentState()
-                                WOO_POS -> skipScreenDuringPosFlow()
+                                WOO_POS -> error("Unsupported card reader flow param: $param")
```
it looks like we can and we should remove this param completely, wdyt?

samiuelson replied:
I'm not sure if we can remove it because we want to mark transactions as made in POS in PaymentIntent metadata, in order to track POS TPV (pffQ75-1aV-p2).

kidinov replied:
Maybe I am missing something, but we build payment intent using code from the `cardreader` module, which you are using via `CardReaderPaymentController`. So, basically, we will need to pass our "POS" option via that `start` and then to  `collectPaymentFlow` method. While `WOO_POS` used to be used only to do different adjustments to the reused IPP flow

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModel.kt:
```diff
@@ -60,9 +106,48 @@ class WooPosTotalsViewModel @Inject constructor(
+
+    private var cardReaderPaymentController: CardReaderPaymentController? = null
+
+    private fun createCardReaderPaymentController(orderId: Long) {
```
I think we could clean this up by extracting this to some factory class that gets all the dependencies via injection and what dynamic gets as parameters

kidinov replied:
Maybe we could make it even injectable in the VM and use the "start" method that accepts `paymentOrRefund` and `isTTPPaymentInProgress`?

samiuelson replied:
> I think we could clean this up by extracting this to some factory class that gets all the dependencies via injection and what dynamic gets as parameters

Good idea. Done in 77984aa13806c6f02ba8369ca0b4e02071136891

> Maybe we could make it even injectable in the VM and use the "start" method that accepts paymentOrRefund and isTTPPaymentInProgress?

I prefer to stick to the initial idea for now, to minimize modifications to the IPP flow as part of this PR.

kidinov reviewed (commented):
Great work @samiuelson ! I left a few comments to think about, please let me know wdyt!

samiuelson committed:
Simplify code

samiuelson committed:
Remove unused code

samiuelson committed:
Extract payment controller creation to factory class

samiuelson commented:
> Great work @samiuelson ! I left a few comments to think about, please let me know wdyt!

Thanks for the thorough review @kidinov  🙌  I addressed the comments. Let me know what you think.

samiuelson committed:
Satisfy detekt's complaints

samiuelson committed:
Fix typo

samiuelson committed:
Merge branch `custom-payment-ui-5` into `custom-payment-ui-6-use-payment-controller-in-pos`

samiuelson committed:
Restrict cancelable states by common interface

To be able to filter them using when expression.

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModel.kt:
```diff
@@ -89,7 +112,16 @@ class WooPosTotalsViewModel @Inject constructor(
             val orderId = dataState.value.orderId
             check(orderId != EMPTY_ORDER_ID)
-            cardReaderFacade.collectPayment(orderId)
+            if (cardReaderFacade.readerStatus.value is Connected) {
```
Just to make sure that it won't merge unnoticed. Now, clicking on the collect payment button won't do anything if a reader is not connected. As [I mentioned here](https://github.com/woocommerce/woocommerce-android/pull/12977/files#r1861052829), we'd still probably need to listen here to the changes in the reader's status and react properly to them.

kidinov reviewed (approved):
@samiuelson Great job!

Approving this as it doesn't go to trunk, but please take a look into my latest comments - I think some of them maybe important at least to be aware of

samiuelson committed:
Attempt to cancel a payment on back press

samiuelson committed:
Revert "Restrict cancelable states by common interface"

This reverts commit a02047d10024b546a79410a8a2445422514cac5b.

samiuelson committed:
Remove "collect payment" button

Start the flow automatically

samiuelson committed:
Use custom, cancelable payment scope

samiuelson committed:
Update paymentScope to inherit from viewModel scope

samiuelson committed:
Remove unused imports

samiuelson committed:
Remove redundant `onClear` implementation

samiuelson committed:
Emit "reader not connected" error in ui state

samiuelson committed:
Add unit test:

given reader not connected, when VM checkout clicked, then should show error

samiuelson committed:
Fix code formatting

samiuelson committed:
Remove redundant val

samiuelson committed:
Add test:

given reader connected, when checkout clicked, then should hide error

samiuelson committed:
Fix typo

samiuelson committed:
Add test:

given reader not connected, when checkout clicked and error CTA clicked, then should try connecting to reader

samiuelson committed:
Build reader not connected error after order created

samiuelson committed:
Render error UI

samiuelson committed:
Update totals screen UI

samiuelson committed:
Clean up code

samiuelson committed:
Collect payment automatically after reader connects

samiuelson committed:
Satisfy detekt's complaints

samiuelson committed:
Add unit test:

given order draft created, when reader connects, then start payment automatically

samiuelson committed:
Add unit tests:

1. given order draft created, when reader connects, then start payment automatically

2. given order draft created, when reader disconnects, then should abort payment action

samiuelson committed:
Clean up code

samiuelson committed:
Clean up code

samiuelson committed:
Update background colors

samiuelson committed:
Show full screen "Payment processing" state

samiuelson committed:
Clean up code

samiuelson committed:
Add preview

samiuelson committed:
Clean up code

samiuelson committed:
Improve preview

samiuelson committed:
Create basic payment failed screen

samiuelson committed:
Show payment failed screen

samiuelson committed:
Update tests

samiuelson committed:
Satisfy detekt's complaints

samiuelson committed:
Tune up payment failed composable

samiuelson committed:
Tune up payment failed composable

samiuelson committed:
Tune up payment failed composable

samiuelson committed:
Rename states

samiuelson committed:
Ensure failed payment state is always full screen

samiuelson committed:
Implement failed payment "retry" action

samiuelson committed:
Implement failed payment "exit order" action

samiuelson committed:
Move debug Totals payment state to the top

samiuelson committed:
Add home VM test:

give home screen is at checkout, when exit order clicked after failed payment, then should show cart

samiuelson committed:
Add home VM test:

give home screen is at checkout, when payment processing started, then should show full screen totals state

samiuelson committed:
Add home VM test:

give home screen is at checkout, processing payment, when payment fails, then should show full screen totals state

samiuelson committed:
Add home VM test:

give home screen is at checkout, failed payment, when retry payment clicked, then should show cart with totals

samiuelson committed:
Tweak heights of totals boxes

samiuelson committed:
Fix typos in test method names

samiuelson committed:
Add Cart VM test:

given non-empty cart, when card payment is aborted, then should clear the cart

samiuelson committed:
Add Totals VM test:

given order draft created and reader connected, when card tapped, should show payment processing screen

samiuelson committed:
Add Totals VM test:

given payment failed, when retry clicked, then should retry

samiuelson committed:
Add Totals VM test:

given payment failed, when retry clicked, then should retry

samiuelson committed:
Add Totals VM test:

given payment failed, when exit order clicked, then should inform home about the situation

samiuelson committed:
Ensure Cart is cleared in case state is restored

samiuelson committed:
Clean up code

samiuelson committed:
Satisfy detekt's complaints

samiuelson committed:
Manage payment coroutine scope internally in controller

samiuelson committed:
Clean up code

samiuelson committed:
Remove unused code

I've checked the history, and `CardReaderPaymentViewModel::reFetchOrder` has never been invoked outside of the class. When extracting payment flow to the controller, I missed the `@VisibleForTesting` annotation and treated it as a public function when it was supposed to be treated as `private`. I'm removing this method from VM – it's supposed to be invoked internally in `CardReaderPaymentController`.

samiuelson committed:
Remove delay(1) from tests

samiuelson committed:
Use WooPosColors for `paymentProcessingBackground`

samiuelson committed:
Use WooPosColors for `totalsBackground`

samiuelson committed:
Replace hardcoded error message with real one

samiuelson committed:
Satisfy detekt's complaints

samiuelson committed:
Get order from db instead of caching in memory

samiuelson committed:
Improve failed payment "retry" handling

samiuelson committed:
Satisfy detekt's complaints

samiuelson committed:
Update string value

Exit order -> Go back to checkout

samiuelson committed:
Handle PaymentCollecting payment state during retry

samiuelson committed:
Handle `PaymentCollecting` payment state received during retry

samiuelson committed:
Clean up code

samiuelson committed:
Implement basic UI for reader initialization states

samiuelson committed:
Tune up the UI

samiuelson committed:
Satisfy detekt's complaints

samiuelson committed:
Revert mistaken rename

samiuelson committed:
Revert mistaken rename

samiuelson committed:
Extract strings to res

samiuelson committed:
Extract strings to res

samiuelson committed:
Fix typo

samiuelson committed:
Update tests

samiuelson committed:
Show subtitles for `processing` and `capturing` states

samiuelson committed:
Show lottie animations for payment capturing and processing

samiuelson committed:
Revert accidental gradle.properties changes

samiuelson committed:
Animate "reader ready" state in Totals

samiuelson committed:
Revert accidental change to gradle.properties

samiuelson committed:
Revert accidental change to gradle.properties

samiuelson committed:
Update unit tests

samiuelson committed:
Add tests:

* given order draft created and reader connected, when payment is processed, should show processing state

* given order draft created and reader connected, when payment is captured, should show processing state

samiuelson committed:
Add payment processing screen enter animation

samiuelson committed:
Satisfy detekt's complaints

samiuelson committed:
Tune up animation

samiuelson committed:
Clean up code

samiuelson committed:
Clean up tests

samiuelson committed:
Remove redundant code

samiuelson committed:
Remove redundant code

samiuelson committed:
Satisfy detekt's complaints

samiuelson committed:
Merge pull request #13037 from woocommerce/custom-payment-ui-7-payment-cancelation

[POS][Custom payment UI] – Payment cancelation

samiuelson committed:
Merge pull request #13038 from woocommerce/custom-payment-iu-remove-collect-payment-button

[POS][Custom payment UI] – Remove "collect payment" button

samiuelson committed:
Give color a name

samiuelson committed:
Merge branch `custom-payment-ui-6-use-payment-controller-in-pos`

samiuelson committed:
Merge pull request #13053 from woocommerce/custom-payment-ui-8-reader-not-connected

[POS][Custom payment UI] – Reader not connected error handling

samiuelson committed:
Merge branch `custom-payment-ui-6-use-payment-controller-in-pos`

samiuelson committed:
Clean up tests after merge

samiuelson committed:
Crash in case order is null illegally

samiuelson committed:
Merge pull request #13057 from woocommerce/custom-payment-ui-9-payment-failed-states

[POS][Custom payment UI] – Failed payment error handling

samiuelson committed:
Merge branch `custom-payment-ui-6-use-payment-controller-in-pos`

samiuelson committed:
Merge pull request #13090 from woocommerce/custom-payment-ui-loading-animation

[POS][Custom payment UI] – UI animation for `Loading` data and `CollectingPayment` states

samiuelson committed:
Merge branch `custom-payment-ui-6-use-payment-controller-in-pos`

samiuelson committed:
Merge branch `trunk`

samiuelson committed:
Merge branch `trunk`

samiuelson committed:
Merge branch `custom-payment-ui-6-use-payment-controller-in-pos`

samiuelson committed:
Rename Processing UI state to InProgress

samiuelson committed:
Merge branch 'trunk' into custom-payment-ui-6-use-payment-controller-in-pos

samiuelson committed:
Merge branch 'custom-payment-ui-6-use-payment-controller-in-pos' into custom-payment-ui-payment-processing-animation

malinajirka committed:
Merge pull request #13100 from woocommerce/custom-payment-ui-payment-processing-animation

[POS][Custom payment UI] – Payment processing animation with Lottie

samiuelson committed:
Remove unused drawable

samiuelson committed:
Update release notes

samiuelson committed:
Extract `isTTPPaymentInProgress` to property delegate

samiuelson committed:
Merge branch 'trunk' into custom-payment-ui-6-use-payment-controller-in-pos

samiuelson committed:
Satisfy detekt's complaints
